src/hyperloglog.c

```diff
@@ -179,29 +179,123 @@
+ *    会认为第 51 位为1，所以 count <= 51
+ * 3) 获取到数据的 index 和 count，将其放入 hyperloglog 数据结构中，具体怎么将数据放入 hyperloglog 需要看
+ * HLL_DENSE_SET_REGISTER 宏
+ *
+ * 四.
+ * 原理和计数（访问的用户量）的计算：
+ * 伯努利过程（简化理解），抛硬币，抛出反面为 0，抛出正面为 1，抛一次两者出现的概率都为（1/2）
+ *    1) 一轮测试，一直抛直到抛出正面或者抛了50次了，本轮测试停止
+ *    （这里一轮测试可以看做一个数据进 hyperloglog 数据结构，该数据的 count 值就是出现正面(1)是第多少次
+ *    每抛一次可以看做这个数据取 hash 之后高50位的一个 bit 位 ，对于用户 ID 相同，hash 值一样，得到的结果一样）
+ *    2) 每次抛出反面的概率都是 1/2，那么第 count 次才出现正面(1) 的概率: P = (1/2)^(count - 1次反面) * (1/2)。
+ *    后面的 1/2 是最后出正面的概率。所以概率就是 P = (1/2)^count
+ *    3) 既然第 count 次才出现正面(1) 的概率是 (1/2)^count。那么理想情况下，做 2^count 轮测试，会出现且仅出现一轮测试
+ *    是第 count 次才出现正面(1)的情况.
+ *    对于大于 count 次才出现正面(1)的情况，理想情况下做 2^count 轮测试不足以出现大于 count 次才出现正面（1）的情况
+ *    （这里 2^count 轮测试可以看做是一个桶的数据量，桶的值就是这 2^count 轮测试抛出最多连续为 0 的那轮测试的值）
+ *    注：到这里可以反推出，理想情况下，如果有且仅有一轮测试出现了第 count 次才出现正面，那么可以认为做了 2^count 轮测试，所以可以根据这个
+ *   最大的 count 值来推算出进来的不同数据的数量是 2^count 次方
+ *    4) 上面获取的结论都是在理想情况下，但是现实中，总是存在偏差，为了减少误差就需要进行分桶，将数据打散到所有桶里面，
+ *    然后用所有桶中的最大 count 值计算平均数 c，
+ *    5) 最后的计算公式： 2^c * 桶的数量 = 用户数量
+ *
+ */
+
 struct hllhdr {
+    /* 定义魔数，HYLL */
     char magic[4];      /* "HYLL" */
+    /* 编码类型，HLL_DENSE 密集型，HLL_SPARSE 稀疏型 */
     uint8_t encoding;   /* HLL_DENSE or HLL_SPARSE. */
     uint8_t notused[3]; /* Reserved for future use, must be zero. */
+    /* 缓存上一次获取计数的值，每次修改 registers 会将该值无效化，获取计数的时候会使用该属性缓存，使用小端字节序存储 */
     uint8_t card[8];    /* Cached cardinality, little endian. */
+    /* 数据字节数组，这里就是存储实际 hyperloglog 数据结构的字节数组 */
     uint8_t registers[]; /* Data bytes. */
 };
 
 /* The cached cardinality MSB is used to signal validity of the cached value. */
+/* 将缓存的计数 card 无效化，因为是小端存储所以是将第8个字节的第一位置 1
+ * 注：如果不熟悉大小端字节序，可以直接看做将 64 位数字的最高位标记为 1
+ */
 #define HLL_INVALIDATE_CACHE(hdr) (hdr)->card[7] |= (1<<7)
+/* 判断 card 是否是有效的，可以直接看做判断最高位是不是 0，如果是 0 就是合法的 */
 #define HLL_VALID_CACHE(hdr) (((hdr)->card[7] & (1<<7)) == 0)
 
+/* 定义将 hash 后的数据低多少位作为索引，该值越大，最后计算计数的误差越小，
+ * 这个用来决定分桶的数量，因为最后会计算平均数（这里不是算术平均数，是调和平均数），如果桶越多，相当于样本越多
+ * 也就是在概率上，用来做实验的样本基数越大，最后的值越趋近于理想值 */
 #define HLL_P 14 /* The greater is P, the smaller the error. */
+/* 高 50 位用来计算这 50 位的低位有多少个连续的 0 */
 #define HLL_Q (64-HLL_P) /* The number of bits of the hash value used for
                             determining the number of leading zeros. */
+/* 桶的数量 */
 #define HLL_REGISTERS (1<<HLL_P) /* With P=14, 16384 registers. */
+/* 桶数量的掩码 */
 #define HLL_P_MASK (HLL_REGISTERS-1) /* Mask to index register. */
+/* 以 6bits 作为一个桶的数据，用来记录当前桶中所有数据起始的连续的 0 的数量 */
 #define HLL_BITS 6 /* Enough to count up to 63 leading zeroes. */
+/* 桶数据的掩码 63 */
 #define HLL_REGISTER_MAX ((1<<HLL_BITS)-1)
+/* hllhdr 结构的头部大小 */
 #define HLL_HDR_SIZE sizeof(struct hllhdr)
+/* 头部加数据部分的总大小， 后面 +7 是因为整数除法会向下取整 */
 #define HLL_DENSE_SIZE (HLL_HDR_SIZE+((HLL_REGISTERS*HLL_BITS+7)/8))
+/* 密集编码 */
 #define HLL_DENSE 0 /* Dense encoding. */
+/* 稀疏编码 */
 #define HLL_SPARSE 1 /* Sparse encoding. */
+/* RAW 编码，仅 redis 内部使用，不会暴露出去 */
```

Comment:
```suggestion
/* RAW 编码，仅 redis 内部使用，不会暴露出去 */
```
不是很懂 hyperloglog，这里指的应该是 hyperloglog 内部使用吧？